apps/src/aichat/redux/aichatRedux.ts

```diff
@@ -395,23 +404,59 @@ export const submitChatContents = createAsyncThunk(
     }
 
     thunkAPI.dispatch(clearChatMessagePending());
-    chatApiResponse.messages.forEach(message =>
-      thunkAPI.dispatch(addChatMessage({...message, timestamp: Date.now()}))
-    );
+    chatApiResponse.messages.forEach(message => {
+      const successfulChatMessage = {...message, timestamp: Date.now()};
+      thunkAPI.dispatch(addChatEvent(successfulChatMessage));
+    });
+  }
+);
+
+// This thunk adds a chat event to chatEventsCurrent (displayed in current chat workspace)
+// if hideForParticipants != true and then logs the event to the backend for all chat events
+// except notifications with includeInHistory != true.
+export const addChatEvent = createAsyncThunk(
+  'aichat/addChatEvent',
+  async (chatEvent: ChatEvent, thunkAPI) => {
+    if (!chatEvent.hideForParticipants) {
+      thunkAPI.dispatch(addEventToChatEventsCurrent(chatEvent));
+    }
+    // Log chat event to backend.
+    const state = thunkAPI.getState() as RootState;
+    const aichatContext: AichatContext = {
+      currentLevelId: parseInt(state.progress.currentLevelId || ''),
+      scriptId: state.progress.scriptId,
+      channelId: state.lab.channel?.id,
+    };
+    // Other than notifications that are not included in chat history (save errors not due to profanity),
+    // log the chat event to backend.
+    if (
+      !isNotification(chatEvent) ||
+      (isNotification(chatEvent) && chatEvent.includeInChatHistory)
+    ) {
+      // If a model update, log only the updated value for temperature and selected model id.
+      // Do not log free text updated values (e.g., system prompt, retrieval contexts, model card info).
+      if (isModelUpdate(chatEvent)) {
+        const {updatedField, updatedValue} = chatEvent;
+        // Only log updated value for temperature and selected model id - free text values are not logged.
+        const updatedValueToLog =
+          updatedField === 'temperature' || updatedField === 'selectedModelId'
+            ? updatedValue
+            : 'N/A';
+        chatEvent = {
+          ...chatEvent,
+          updatedValue: updatedValueToLog,
+        } as ModelUpdate;
```

Comment:
Same here, why is the cast needed? Just curious because I would have expected that the `isModelUpdate` function would have already narrowed the type

Reply:
Without cast: `Object literal may only specify known properties, and 'updatedValue' does not exist in type 'ChatEvent'.ts(2353)`